packages/fxa-settings/src/pages/Index/container.tsx

```diff
@@ -0,0 +1,157 @@
+import { AuthUiError, AuthUiErrors } from '../../lib/auth-errors/auth-errors';
+import { getHandledError } from '../../lib/error-utils';
+import { currentAccount } from '../../lib/cache';
+import { useValidatedQueryParams } from '../../lib/hooks/useValidate';
+import { IndexQueryParams } from '../../models/pages/index';
+import { isUnsupportedContext } from '../../models/integrations/utils';
+import { hardNavigate } from 'fxa-react/lib/utils';
+import LoadingSpinner from 'fxa-react/components/LoadingSpinner';
+import { checkEmailDomain } from '../../lib/email-domain-validator';
+import { isEmailMask, isEmailValid } from 'fxa-shared/email/helpers';
+
+// TODO: remove this function, it's only here to make TS happy until
+// we work on FXA-9757. errnos are always defined
+function getErrorWithDefinedErrNo(error: AuthUiError) {
+  return {
+    ...error,
+    errno: error.errno!,
+  };
+}
+
+export const IndexContainer = ({
+  integration,
+  serviceName,
+}: IndexContainerProps & RouteComponentProps) => {
+  // TODO, more strict validation for bad oauth params, FXA-11297
+  const authClient = useAuthClient();
+  const navigate = useNavigate();
+  const location = useLocation() as ReturnType<typeof useLocation> & {
+    state?: LocationState;
+  };
+  const { queryParamModel, validationError } =
+    useValidatedQueryParams(IndexQueryParams);
+
+  const { prefillEmail, deleteAccountSuccess, hasBounced } =
+    location.state || {};
+
+  const isWebChannelIntegration =
+    integration.isSync() || integration.isDesktopRelay();
+
+  // Query param should take precedence
+  const email = queryParamModel.email || currentAccount()?.email;
+  const shouldRedirectToSignin = email && !prefillEmail;
+
+  useEffect(() => {
+    if (shouldRedirectToSignin) {
+      navigate('/signin', {
+        state: {
+          email,
+        },
+      });
+    }
+  });
+
+  const signUpOrSignInHandler = useCallback(
+    async (email: string) => {
+      try {
+        if (!isEmailValid(email)) {
+          return {
+            error: getErrorWithDefinedErrNo(AuthUiErrors.EMAIL_REQUIRED),
+          };
+        }
+        const { exists, hasLinkedAccount, hasPassword } =
+          await authClient.accountStatusByEmail(email, {
+            thirdPartyAuthStatus: true,
+          });
+
+        if (!exists) {
+          if (isEmailMask(email)) {
+            return {
+              error: getErrorWithDefinedErrNo(
+                AuthUiErrors.EMAIL_MASK_NEW_ACCOUNT
+              ),
+            };
+            // "@firefox" or "@firefox.com" email addresses are not valid
+            // at this time, therefore block the attempt.
+            // the added 'i' disallows uppercase letters
+          } else if (new RegExp('@firefox(\\.com)?$', 'i').test(email)) {
+            return {
+              error: getErrorWithDefinedErrNo(
+                AuthUiErrors.DIFFERENT_EMAIL_REQUIRED_FIREFOX_DOMAIN
+              ),
+            };
```

Comment:
Shouldn't these checks also happen on the server side? I can still replay my request with the server and create an account with it right? Probably follow-up work since this is how it was implemented in Backbone too from what I read and this patch just surfaces it.

Reply:
Yep good call out, there's a related comment here: https://github.com/mozilla/fxa/pull/18498#discussion_r1989523898